From 46c16e5f6dd6175ad19d0b0147bd6a69304c2072 Mon Sep 17 00:00:00 2001
From: Paul Moore <pmoore@redhat.com>
Date: Tue, 18 Nov 2014 23:15:13 -0500
Subject: [kernel] audit: keep inode pinned

Message-id: <20141118231513.5995.33120.stgit@localhost>
Patchwork-id: 100269
O-Subject: [RHEL7 PATCH] audit: keep inode pinned
Bugzilla: 1162261
RH-Acked-by: Richard Guy Briggs <rbriggs@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1162261
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8256751
Upstream: yes, see commit ID below
Tested: yes, see BZ for details and reproducer

 commit 799b601451b21ebe7af0e6e8f6e2ccd4683c5064
 Author: Miklos Szeredi <mszeredi@suse.cz>
 Date:   Tue Nov 4 11:27:12 2014 +0100

 audit: keep inode pinned

 Audit rules disappear when an inode they watch is evicted from the cache.
 This is likely not what we want.

 The guilty commit is "fsnotify: allow marks to not pin inodes in core",
 which didn't take into account that audit_tree adds watches with a zero
 mask.

 Adding any mask should fix this.

 Fixes: 90b1e7a57880 ("fsnotify: allow marks to not pin inodes in core")
 Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
 Cc: stable@vger.kernel.org # 2.6.36+
 Signed-off-by: Paul Moore <pmoore@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/audit_tree.c b/kernel/audit_tree.c
index 43c307d..00c4459 100644
--- a/kernel/audit_tree.c
+++ b/kernel/audit_tree.c
@@ -154,6 +154,7 @@ static struct audit_chunk *alloc_chunk(int count)
   chunk->owners[i].index = i;
  }
  fsnotify_init_mark(&chunk->mark, audit_tree_destroy_watch);
+ chunk->mark.mask = FS_IN_IGNORED;
  return chunk;
 }
 
-- 
1.7.1